itou/www/invitations_views/views.py

```diff
@@ -53,28 +73,16 @@ def new_user(request, invitation_type, invitation_id, template_name="invitations
                 url=login_url,
                 redirect_to=get_safe_url(request, "redirect_to"),
             )
-            next_step = redirect(next_step_url)
+            return redirect(next_step_url)
         else:
             # A new user should be created before joining
-            if invitation_type == KIND_PRESCRIBER:
-                params = {
-                    "user_kind": KIND_PRESCRIBER,
-                    "previous_url": request.path_info,
-                    "next_url": get_safe_url(request, "redirect_to"),
-                }
-                inclusion_connect_url = f"{reverse('inclusion_connect:authorize')}?{urlencode(params)}"
-                context["inclusion_connect_url"] = inclusion_connect_url
-                return render(request, "invitations_views/new_prescriber.html", context=context)
-            form = NewUserInvitationForm(data=request.POST or None, invitation=invitation)
-            context["form"] = form
-            if form.is_valid():
-                user = form.save(request)
-                get_adapter().login(request, user)
-                next_step = redirect(get_safe_url(request, "redirect_to"))
+            handle_registration = {
+                KIND_PRESCRIBER: handle_invited_prescriber_registration,
+            }.get(invitation_type, handle_invited_prescriber_registration)
+            return handle_registration(request, context)
     else:
         messages.error(request, "Cette invitation n'est plus valide.")
-
-    return next_step or render(request, template_name, context=context)
+        return render(request, "invitations_views/invitation_errors.html", context=context)
```

Comment:
j'aurais mis du coup tout le else en haut en mode barrage:
```
def new_user(request, invitation_type, invitation_id):
    [...]
    if not request.user.is_authenticated:
        return render(invitation_errors)

    if request.user.email == invitation.email:
        # je note que `invitation_type` est en paramètre alors que c'est censé etre `request.user.kind` si cette prop existait :/
        # je note aussi que le "context" c'est juste l'invitation, y a moyen de simplifier.
        return registration_backend(invitation_type)(request, context)

    return redirect(next_url)
```

joli refactor !

et du coup le `handle

Reply:
`invitation_errors` ce n'est pas en rapport avec l'authentification mais si l'invitation n'est pas valide.

En gros : 
1) si authentifié avec un mauvais utilisateur -> NOGO
2) si invitation valide
 a) si le user existe -> redirection vers login suivie de la vue d'ajout à l'organisation. si déjà authentifié la vue de login est un passe-plat
 b) sinon (pas de user) on arrive sur la page de création (django ou IC)
3) si invitation non valide -> on affiche les erreurs

Je peux donc inverser 3) et 2) pour mettre 
```
if not invitation.can_be_accepted:
    messages.error(request, "Cette invitation n'est plus valide.")
    return render(request, "invitations_views/invitation_errors.html", context=context)
   
user = User.objects.filter(email__iexact=invitation.email)
# bla bla bla
``` 

Et c'est noté pour le context !